src/ClaimerAgent.sol

```diff
@@ -4,76 +4,219 @@ pragma solidity 0.8.17;
+        uint256 prizeCount;
+        bool init = false;
+        uint8 tier;
+        for (uint p = _nextPrizeIndex; p < _nextPrizeIndex + _claimCount; p++) {
+            if (!init) {
+                tier = drawPrizes[computedDrawId][p].tier;
+                init = true;
+            }
+            if (tier != drawPrizes[computedDrawId][p].tier) {
+                break;
+            }
+            prizeCount++;
+        }
+        return (tier, prizeCount);
+    }
+
+    function countWinners(uint _nextPrizeIndex, uint _claimCount) public view returns (uint256 count) {
+        uint256 winnersLength;
+        address lastWinner;
+        for (uint p = _nextPrizeIndex; p < _nextPrizeIndex + _claimCount; p++) {
+            if (drawPrizes[computedDrawId][p].winner != lastWinner) {
+                winnersLength++;
+                lastWinner = drawPrizes[computedDrawId][p].winner;
+            }
+        }
+        return winnersLength;
+    }
+
+    function countPrizeIndicesPerWinner(
+        uint _nextPrizeIndex,
+        uint _claimCount,
+        uint _winnersLength
+    ) public view returns (uint32[] memory prizeIndexLength) {
+        uint256 winnersLength = countWinners(_nextPrizeIndex, _claimCount);
+        prizeIndexLength = new uint32[](_winnersLength);
+        uint256 winnerIndex;
+        address currentWinner;
+        for (uint p = _nextPrizeIndex; p < _nextPrizeIndex + _claimCount; p++) {
+            if (currentWinner == address(0)) {
+                currentWinner = drawPrizes[computedDrawId][p].winner;
+            }
+            if (currentWinner != drawPrizes[computedDrawId][p].winner) {
+                winnerIndex++;
+                currentWinner = drawPrizes[computedDrawId][p].winner;
+            }
+            prizeIndexLength[winnerIndex]++;
+        }
+    }
+
+    function populateArrays(
+        uint _nextPrizeIndex,
+        uint _claimCount,
+        uint _winnersLength,
+        uint32[] memory _prizeIndexLength
+    ) public view returns (address[] memory winners, uint32[][] memory prizeIndices) {
+        winners = new address[](_winnersLength);
+        prizeIndices = new uint32[][](_winnersLength);
+        for (uint i = 0; i < _winnersLength; i++) {
+            prizeIndices[i] = new uint32[](_prizeIndexLength[i]);
+        }
+
+        // populate result arrays
+        uint winnerIndex = 0;
+        address currentWinner;
+        uint256 prizeIndex;
+        for (uint pi = _nextPrizeIndex; pi < _nextPrizeIndex + _claimCount; pi++) {
+            if (currentWinner == address(0)) {
+                currentWinner = drawPrizes[computedDrawId][pi].winner;
+                winners[winnerIndex] = currentWinner;
             }
-            env.claimer().claimPrizes(drawId, claims, address(this));
-            console2.log("CLAIMING ", claimCount, " PRIZES");
-            draws[drawId].currentClaimIndex += claimCount;
+
+            // if the current prize is for a different winner, then skip ahead
```

Comment:
This comment is a bit misleading.
If the `currentWinner` is different from the current prize winner, we don't skip ahead but enter in the condition and update `currentWinner` with this new winner.
So maybe it should be rephrased `then skip to the next winner`.